### ISSUE TITLE
Plugin hangs on start if it starts before docker

### ISSUE THREAD

**Issue by davemuench** [vmware/docker-volume-vsphere#664]
Environment Details: ESXi 6.0u2 single host with iSCSI storage, Docker 1.12.3 swarm on Ubuntu 16.04.1 VMs (3 manager nodes, 3 worker nodes)

Steps to Reproduce:
1. Install the 0.7 or 0.8 plugin
2. Reboot the VM
3. The plugin hangs on trying to read the docker socket.
4. If you restart the plugin at that point, it loads correctly.

If you modify the startup dependencies so docker loads before the plugin, it works fine.

Expected Result:

Normal startup as seen in this log file:

2016-10-29 10:17:44.2543284 -0700 PDT [INFO] No config file found. Using defaults.
2016-10-29 10:17:44.254428311 -0700 PDT [INFO] Docker VMDK plugin started config="/etc/docker-volume-vsphere.conf" version="VMDK Volume Driver v0.3" port=1019 mock_esx=false log_level=info
2016-10-29 10:17:44.254463997 -0700 PDT [INFO] Getting volume data from unix:///var/run/docker.sock
2016-10-29 10:17:44.261199728 -0700 PDT [INFO] Discovered 0 volumes in use.
2016-10-29 10:17:44.261246076 -0700 PDT [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock"

Actual Result:

The log file stops here:

2016-10-20 17:13:58.834392451 -0700 PDT [INFO] No config file found. Using defaults.
2016-10-20 17:13:58.836004394 -0700 PDT [INFO] Docker VMDK plugin started log_level=info config="/etc/docker-volume-vsphere.conf" version="VMDK Volume Driver v0.3" port=1019 mock_esx=false
2016-10-20 17:13:58.836101748 -0700 PDT [INFO] Getting volume data from unix:///var/run/docker.sock

After sending a restart to the plugin, the log file resumes here:

2016-10-28 13:26:50.541774514 -0700 PDT [INFO] Can't connect to unix:///var/run/docker.sock, skipping discovery
2016-10-28 13:26:50.541964669 -0700 PDT [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock"
2016-10-28 13:26:50.543130006 -0700 PDT [WARNING] Received signal signal=terminated
2016-10-29 10:17:44.2543284 -0700 PDT [INFO] No config file found. Using defaults.
2016-10-29 10:17:44.254428311 -0700 PDT [INFO] Docker VMDK plugin started config="/etc/docker-volume-vsphere.conf" version="VMDK Volume Driver v0.3" port=1019 mock_esx=false log_level=info
2016-10-29 10:17:44.254463997 -0700 PDT [INFO] Getting volume data from unix:///var/run/docker.sock
2016-10-29 10:17:44.261199728 -0700 PDT [INFO] Discovered 0 volumes in use.
2016-10-29 10:17:44.261246076 -0700 PDT [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock"

You can see the plugin on that node was hung for a week until I restarted it. I upgraded to 0.8 today and the problem remains. Apologies for not reporting it back on 0.7, but 0.7 was broken in other ways for me that 0.8 fixed.

**Comment by pdhamdhere** [comment]
```	
log.Infof("Getting volume data from %s", dockerUSocket)
info, err := c.Info(context.Background())  <<<< **Docker never returns?**
if err != nil {
	log.Infof("Can't connect to %s, skipping discovery", dockerUSocket)
```
@davemuench Can you please share Docker Engine log /var/log/upstart/docker.log and output of `docker info`.

CC / @kerneltime @msterin

**Comment by davemuench** [comment]
Here's some fresh logs after I rebooted this particular worker node. The docker engine logs are from journalctl since this is Ubuntu 16.04.

docker-volume-vsphere.log:

2016-10-30 11:33:02.383347395 -0700 PDT [INFO] No config file found. Using defaults.
2016-10-30 11:33:02.386559429 -0700 PDT [INFO] Docker VMDK plugin started log_level=info config="/etc/docker-volume-vsphere.conf" version="VMDK Volume Driver v0.3" port=1019 mock_esx=false
2016-10-30 11:33:02.386665908 -0700 PDT [INFO] Getting volume data from unix:///var/run/docker.sock

< This is where I restarted the plugin service >

2016-10-30 11:34:35.106478481 -0700 PDT [WARNING] Received signal signal=terminated
2016-10-30 11:34:35.126785096 -0700 PDT [INFO] No config file found. Using defaults.
2016-10-30 11:34:35.126844689 -0700 PDT [INFO] Docker VMDK plugin started mock_esx=false log_level=info config="/etc/docker-volume-vsphere.conf" version="VMDK Volume Driver v0.3" port=1019
2016-10-30 11:34:35.126920388 -0700 PDT [INFO] Getting volume data from unix:///var/run/docker.sock
2016-10-30 11:34:35.148208031 -0700 PDT [INFO] Discovered 0 volumes in use.
2016-10-30 11:34:35.148244487 -0700 PDT [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock"

journalctl output for docker.service from the same time period:

Oct 30 11:33:02 docker3 systemd[1]: Starting Docker Application Container Engine...
Oct 30 11:33:02 docker3 dockerd[1271]: time="2016-10-30T11:33:02.828293724-07:00" level=info msg="libcontainerd: new containerd process, pid: 1316"
Oct 30 11:33:03 docker3 dockerd[1271]: time="2016-10-30T11:33:03.943726553-07:00" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Oct 30 11:33:03 docker3 dockerd[1271]: time="2016-10-30T11:33:03.944358796-07:00" level=info msg="Loading containers: start."
Oct 30 11:33:03 docker3 dockerd[1271]: .....time="2016-10-30T11:33:03.994583695-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.133585877-07:00" level=info msg="Removing stale sandbox ingress_sbox (ingress-sbox)"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.178186513-07:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.213118167-07:00" level=info msg="Loading containers: done."
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.275057247-07:00" level=info msg="Initializing Libnetwork Agent Listen-Addr=0.0.0.0 Local-addr=10.0.11.63 Adv-addr=10.0.11.63 Remote-addr =10.0.11.52"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.275348966-07:00" level=info msg="Gossip cluster hostname docker3-300bf31dbddd"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.275133303-07:00" level=info msg="Daemon has completed initialization"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.275496725-07:00" level=info msg="Docker daemon" commit=6b644ec graphdriver=overlay version=1.12.3
Oct 30 11:33:04 docker3 systemd[1]: Started Docker Application Container Engine.
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.290560230-07:00" level=info msg="API listen on /var/run/docker.sock"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04.290634949-07:00" level=info msg="API listen on [::]:2376"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:04 docker3 dockerd[1271]: time="2016-10-30T11:33:04-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:34 docker3 dockerd[1271]: time="2016-10-30T11:33:34.439807378-07:00" level=error msg="Bulk sync to node docker2-0c10810a471c timed out"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"
Oct 30 11:33:36 docker3 dockerd[1271]: time="2016-10-30T11:33:36-07:00" level=info msg="Firewalld running: false"

docker info output:

Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 5
Server Version: 1.12.3
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: active
 NodeID: 0xfdzbltn2y6v024ov7gf3sgg
 Is Manager: false
 Node Address: 10.0.11.63
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-45-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: docker3
ID: GGHK:JMML:TS2F:AXNX:3J3C:46SA:NRG5:HWYR:2FVB:S326:LMNZ:RQM5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8

Let me know if you need anything else. Thanks.

**Comment by govint** [comment]
@davemuench can you also provide the logs from /var/log/upstarts/docker.log for the same time period that the plugin was hung.

**Comment by davemuench** [comment]
@govint As I mentioned this is Ubuntu 16.04 - upstart no longer exists, and that logfile doesn't exist. The journalctl output above should be the equivalent and was for the time period the plugin was hung (11:33:02 - 11:34:35).

If there is some other log I am unaware of please let me know but I believe that is it.

**Comment by pdhamdhere** [comment]
Nothing's stands out from logs. We've Ubuntu 14.04 is CI. Will try to reproduce with 16.04.

**Comment by kerneltime** [comment]
@davemuench as a datapoint I was not able to reproduce this on Photon OS running an version of Docker (1.12.1)
```
root@photon-3Q686zl9Z [ ~ ]# systemctl start docker-volume-vsphere
2016-10-31 05:24:52.932938727 +0000 UTC [INFO] No config file found. Using defaults.
2016-10-31 05:24:52.935188294 +0000 UTC [INFO] Docker VMDK plugin started version="VMDK Volume Driver v0.3" port=1019 mock_esx=false log_level=info config="/etc/docker-volume-vsphere.conf" 
2016-10-31 05:24:52.935600317 +0000 UTC [INFO] Getting volume data from unix:///var/run/docker.sock
2016-10-31 05:24:52.93604613 +0000 UTC [INFO] Can't connect to unix:///var/run/docker.sock, skipping discovery
2016-10-31 05:24:52.936113502 +0000 UTC [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock" 
root@photon-3Q686zl9Z [ ~ ]# systemctl start docker
```

**Comment by pdhamdhere** [comment]
@davemuench I tried with Ubuntu 16.04.1 LTS but no luck in reproducing the issue. Tested with Swarm mode on both master and slave too and everything worked like a charm :(

However, I did notice on my setup Docker starts before plugin. In your case, Docker is not running and for some reason call to Docker [`c.Info(context.Background())`](https://godoc.org/github.com/docker/engine-api/client#Client.Info) never returns. 

I did force my setup to start plugin before docker and it still worked fine.

**Comment by kerneltime** [comment]
@davemuench can you email plot.svg to cna-storage@vmware.com after the following command
```
systemd-analyze plot > plot.svg
```

**Comment by govint** [comment]
One difference is who listens on /var/run/doocker.sock in 16.04. In 14.04 docker listens on this socket but in the new release its init process that listens on the socket.
  ---
14.04

unix  2      [ ACC ]     STREAM     LISTENING     491232   31719/dockerd       /var/run/docker.sock <-----------
unix  2      [ ACC ]     STREAM     LISTENING     491242   31736/docker-contai /var/run/docker/libcontainerd/docker-containerd.sock
unix  2      [ ACC ]     STREAM     LISTENING     10740    1277/nslcd          /var/run/nslcd/socket
unix  2      [ ACC ]     STREAM     LISTENING     499964   1163/docker-volume- /run/docker/plugins/vmdk.sock


and 16.04,

unix  2      [ ACC ]     STREAM     LISTENING     79568    13374/docker-contai /var/run/docker/libcontainerd/docker-containerd.sock
unix  2      [ ACC ]     STREAM     LISTENING     79736    13369/dockerd       /run/docker/libnetwork/97c7b0a490267e1a56d6aa71ca60ac9a6e3fca49e32442a95d7705ccaae41b66.sock
unix  2      [ ACC ]     STREAM     LISTENING     81466    13837/docker-volume /run/docker/plugins/vmdk.sock
unix  2      [ ACC ]     STREAM     LISTENING     21332    3889/dbus-daemon    @/tmp/dbus-jF4EVplrqT
unix  3      [ ]         DGRAM                    10454    1/init              /run/systemd/notify
unix  2      [ ACC ]     STREAM     LISTENING     79424    1/init              /var/run/docker.sock <-------------------

**Comment by davemuench** [comment]
I've emailed the plot output to that address as requested.

FWIW, I've worked around the issue by putting a small sleep in /etc/rc.local followed by restarting the docker-volume-vsphere service. Ugly, but a bit easier than changing the systemd dependencies to have the plugin start second.

I do have a systemd override in place for the docker engine to enable the overlay storage driver and turn on TLS:

[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock -s overlay --tlsverify --tlscacert=/path/ca.pem --tlscert=/path/%H-cert.pem --tlskey=/path/%H-key.pem

Other than that I haven't modified the docker (or plugin) startup order or unit files.

**Comment by ashahi1** [comment]
I tried with following setup but was not able to reproduce the issue.
Setup: Ubuntu 16.04.1 LTS - Docker 1.12.3 - 0.8 Plugin

Also tested with swarm mode but things worked without any issues.

**Comment by shuklanirdesh82** [comment]
closing this as per previous comment. Thanks @ashahi1 !

**Comment by tusharnt** [comment]
@pdhamdhere mentioned that he was running into this scenario recently. I will let him decide whether to close this issue.

**Comment by tusharnt** [comment]
@msterin would this be addressed by the fix for #886 ?

**Comment by msterin** [comment]
> @msterin would this be addressed by the fix for #886 ?

not in Docker 1.13. Managed plugin support is experimental in docker and they do not recommend it for production, at least until they announce full support. 

Also 
> @pdhamdhere mentioned that he was running into this scenario recently. I will let him decide whether to close this issue.

Please do not reopen issues without supporting evidence - a repro or at least logs, config  and details. Otherwise I do not see what we can do about it, and it's not a blocker anyways. Since there is no action to be taken  - closing it. 

**If and when** we do have a repro, feel free to reopen
